# Let's Stay Up-to-Date!

With the project completed during the end of 2020 and with COVID running rampant, altering the timeline of the season as well as turning MLB around, I feel like it is time to check and see if the umpires have gotten better since I last analyzed their performances.

In [1]:
# Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import pybaseball as pb
from pybaseball import cache

In [2]:
pd.set_option('display.max_rows', None, 'display.max_columns', None)

In [3]:
cache.enable()

## Let's Retrieve New Data (and try to use the old one we already have)

In [4]:
# The last data retrieved was the 2020 MLB season where the LA Dodgers won, now we have a new champion the Atlanta Braves
# Previous data have already been saved in a different folder

# mlb2021 = pb.statcast(start_dt='2021-04-01', end_dt='2021-11-02')
# print('Retrieval Complete')

This is a large query, it may take a moment to complete
Completed sub-query from 2021-04-01 to 2021-04-06
Completed sub-query from 2021-04-07 to 2021-04-12
Completed sub-query from 2021-04-13 to 2021-04-18
Completed sub-query from 2021-04-19 to 2021-04-24
Completed sub-query from 2021-04-25 to 2021-04-30
Completed sub-query from 2021-05-01 to 2021-05-06
Completed sub-query from 2021-05-07 to 2021-05-12
Completed sub-query from 2021-05-13 to 2021-05-18
Completed sub-query from 2021-05-19 to 2021-05-24
Completed sub-query from 2021-05-25 to 2021-05-30
Completed sub-query from 2021-05-31 to 2021-06-05
Completed sub-query from 2021-06-06 to 2021-06-11
Completed sub-query from 2021-06-12 to 2021-06-17
Completed sub-query from 2021-06-18 to 2021-06-23
Completed sub-query from 2021-06-24 to 2021-06-29
Completed sub-query from 2021-06-30 to 2021-07-05
Completed sub-query from 2021-07-06 to 2021-07-11
Completed sub-query from 2021-07-12 to 2021-07-17
Completed sub-query from 2021-07-18 to 2021-

In [5]:
# Save this to a csv file in case I need to recall it.
# mlb2021.to_csv('2021mlb.csv', index=False)

In [4]:
# Skip retrieval since we already retrieved
# Read the CSV

mlb2021 = pd.read_csv('2021mlb.csv')

Let's look at the data to make sure it has all the correct information.

To get an understanding of what these columns stand for, [click here](https://baseballsavant.mlb.com/csv-docs).

In [5]:
mlb2021.head(5) # We have the last pitches of the World Series

index pitch_type   game_date  release_speed  release_pos_x  release_pos_z  \
0    114         FF  2021-11-02           93.7           1.39           6.72   
1    117         FF  2021-11-02           92.9           1.38           6.72   
2    128         FF  2021-11-02           93.1           1.35           6.73   
3    130         FF  2021-11-02           94.6           1.31           6.73   
4    140         FF  2021-11-02           93.6           1.31           6.80   

   player_name    batter   pitcher     events    description  spin_dir  \
0  Smith, Will  493329.0  519293.0  field_out  hit_into_play       NaN   
1  Smith, Will  493329.0  519293.0        NaN           foul       NaN   
2  Smith, Will  493329.0  519293.0        NaN  called_strike       NaN   
3  Smith, Will  670541.0  519293.0  field_out  hit_into_play       NaN   
4  Smith, Will  670541.0  519293.0        NaN           ball       NaN   

   spin_rate_deprecated  break_angle_deprecated  break_length_deprecated  \
0                   NaN                     NaN                      NaN   
1                   NaN                     NaN                      NaN   
2                   NaN                     NaN                      NaN   
3                   NaN                     NaN                      NaN   
4                   NaN                     NaN                      NaN   

   zone                                                des game_type stand  \
0   5.0  Yuli Gurriel grounds out, shortstop Dansby Swa...         W     R   
1   5.0  Yuli Gurriel grounds out, shortstop Dansby Swa...         W     R   
2   6.0  Yuli Gurriel grounds out, shortstop Dansby Swa...         W     R   
3   5.0  Yordan Alvarez flies out to left fielder Eddie...         W     L   
4  13.0  Yordan Alvarez flies out to left fielder Eddie...         W     L   

  p_throws home_team away_team type  hit_location      bb_type  balls  \
0        L       HOU       ATL    X           6.0  ground_ball    0.0   
1        L       HOU       ATL    S           NaN          NaN    0.0   
2        L       HOU       ATL    S           NaN          NaN    0.0   
3        L       HOU       ATL    X           7.0     fly_ball    3.0   
4        L       HOU       ATL    B           NaN          NaN    2.0   

   strikes  game_year  pfx_x  pfx_z  plate_x  plate_z  on_3b  on_2b     on_1b  \
0      2.0     2021.0   0.57   1.21     0.19     2.68    NaN    NaN  488726.0   
1      1.0     2021.0   0.90   1.34     0.01     2.12    NaN    NaN  488726.0   
2      0.0     2021.0   0.81   1.52     0.78     2.13    NaN    NaN  488726.0   
3      2.0     2021.0   0.85   1.27    -0.23     2.66    NaN    NaN  488726.0   
4      2.0     2021.0   0.90   1.43    -1.15     1.51    NaN    NaN  488726.0   

   outs_when_up  inning inning_topbot   hc_x    hc_y  tfs_deprecated  \
0           2.0     9.0           Bot  98.12  136.43             NaN   
1           2.0     9.0           Bot    NaN     NaN             NaN   
2           2.0     9.0           Bot    NaN     NaN             NaN   
3           1.0     9.0           Bot  68.50   86.79             NaN   
4           1.0     9.0           Bot    NaN     NaN             NaN   

   tfs_zulu_deprecated  fielder_2  umpire  sv_id       vx0         vy0  \
0                  NaN   518595.0     NaN    NaN -4.348927 -136.263269   
1                  NaN   518595.0     NaN    NaN -5.455433 -134.989926   
2                  NaN   518595.0     NaN    NaN -3.230974 -135.201801   
3                  NaN   518595.0     NaN    NaN -5.901934 -137.422092   
4                  NaN   518595.0     NaN    NaN -8.265487 -135.578023   

         vz0         ax         ay         az  sz_top  sz_bot  \
0  -7.335202   8.101619  29.384843 -15.743652    3.37    1.53   
1  -8.881689  12.188852  30.690010 -14.009571    3.37    1.53   
2  -9.255781  10.678480  31.699974 -11.705800    3.40    1.53   
3  -7.652311  12.118242  35.102245 -14.577857    3.58    1.68   
4 -10.936295  12.93280

In [6]:
mlb2021.tail(3) # And it confirms the first games of the year when cross referencing baseball-reference.com

index pitch_type   game_date  release_speed  release_pos_x  \
720634  22008         FF  2021-04-01           96.1          -1.76   
720635  22073         FF  2021-04-01           95.6          -2.04   
720636  22343         FF  2021-04-01           95.1          -1.95   

        release_pos_z      player_name    batter   pitcher events description  \
720634           6.27  Márquez, Germán  605141.0  608566.0    NaN        ball   
720635           6.03  Márquez, Germán  605141.0  608566.0    NaN        foul   
720636           6.11  Márquez, Germán  605141.0  608566.0    NaN        ball   

        spin_dir  spin_rate_deprecated  break_angle_deprecated  \
720634       NaN                   NaN                     NaN   
720635       NaN                   NaN                     NaN   
720636       NaN                   NaN                     NaN   

        break_length_deprecated  zone  \
720634                      NaN  11.0   
720635                      NaN   5.0   
720636                      NaN  12.0   

                                                      des game_type stand  \
720634  Mookie Betts singles on a fly ball to left fie...         R     R   
720635  Mookie Betts singles on a fly ball to left fie...         R     R   
720636  Mookie Betts singles on a fly ball to left fie...         R     R   

       p_throws home_team away_team type  hit_location bb_type  balls  \
720634        R       COL       LAD    B           NaN     NaN    1.0   
720635        R       COL       LAD    S           NaN     NaN    1.0   
720636        R       COL       LAD    B           NaN     NaN    0.0   

        strikes  game_year  pfx_x  pfx_z  plate_x  plate_z  on_3b  on_2b  \
720634      1.0     2021.0  -0.42   0.86    -0.61     3.65    NaN    NaN   
720635      0.0     2021.0  -0.25   0.88     0.19     2.60    NaN    NaN   
720636      0.0     2021.0  -0.30   1.01     0.07     4.16    NaN    NaN   

        on_1b  outs_when_up  inning inning_topbot  hc_x  hc_y  tfs_deprecated  \
720634    NaN           0.0     1.0           Top   NaN   NaN             NaN   
720635    NaN           0.0     1.0           Top   NaN   NaN             NaN   
720636    NaN           0.0     1.0           Top   NaN   NaN             NaN   

        tfs_zulu_deprecated  fielder_2  umpire  sv_id       vx0         vy0  \
720634                  NaN   553869.0     NaN    NaN  3.953292 -140.092280   
720635                  NaN   553869.0     NaN    NaN  6.352185 -139.293012   
720636                  NaN   553869.0     NaN    NaN  5.887082 -138.617211   

             vz0        ax         ay         az  sz_top  sz_bot  \
720634 -3.371124 -6.331036  25.767684 -20.177597    3.34    1.64   
720635 -5.428763 -4.405496  25.441128 -19.766641    3.29    1.49   
720636 -1.813448 -4.967683  25.527820 -18.815090    3.46    1.76   

        hit_distance_sc  launch_speed  launch_angle  effective_speed  \
720634              NaN           NaN           NaN             95.6   
720635            265.0          77.7          39.0             95.1   
720636              NaN           NaN           NaN             94.5   

        release_spin_rate  release_extension   game_pk  pitcher.1  \
720634             2063.0                5.5  634615.0   608566.0   
720635             2988.0                5.5  634615.0   608566.0   
720636                NaN                5.5  634615.0   608566.0   

        fielder_2.1  fielder_3  fielder_4  fielder_5  fielder_6  fielder_7  \
720634     553869.0   543068.0   572008.0   658069.0   596115.0   606132.0   
720635     553869.0   543068.0   572008.0   658069.0   596115.0   606132.0   
720636     553869.0   543068.0   572008.0   658069.0   596115.0   606132.0   

        fielder_8  fielder_9  release_pos_y  estimated_ba_using_speedangle  \
720634   641658.0   453568.0          55.01                            NaN   
720635   641658.0   453568.0          54.98                            NaN   
720636   641658.0   453568.0         

# Next Steps

In order for me to be able to work with the data, it needs to go through some rigourous cleaning. There are a lot of columns that are not helpful in determining if I can run a model on this but it should be possible since we are only focused on a few things.

## Understanding the data

In [7]:
mlb2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 720637 entries, 0 to 720636
Data columns (total 93 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   index                            720637 non-null  int64  
 1   pitch_type                       720272 non-null  object 
 2   game_date                        720637 non-null  object 
 3   release_speed                    720236 non-null  float64
 4   release_pos_x                    720059 non-null  float64
 5   release_pos_z                    720059 non-null  float64
 6   player_name                      720637 non-null  object 
 7   batter                           720637 non-null  float64
 8   pitcher                          720637 non-null  float64
 9   events                           184107 non-null  object 
 10  description                      720637 non-null  object 
 11  spin_dir                         0 non-null       float64
 12  sp

In [8]:
mlb2021.isnull().sum()

index                                   0
pitch_type                            365
game_date                               0
release_speed                         401
release_pos_x                         578
release_pos_z                         578
player_name                             0
batter                                  0
pitcher                                 0
events                             536530
description                             0
spin_dir                           720637
spin_rate_deprecated               720637
break_angle_deprecated             720637
break_length_deprecated            720637
zone                                  368
des                                     0
game_type                               0
stand                                   0
p_throws                                0
home_team                               0
away_team                               0
type                                    0
hit_location                      

In [9]:
mlb2021.describe()

index  release_speed  release_pos_x  release_pos_z  \
count  720637.000000  720236.000000  720059.000000  720059.000000   
mean    11465.853512      88.863055      -0.702066       5.855274   
std      6869.474401       6.053327       1.889028       0.529734   
min         0.000000      30.100000      -5.340000       0.940000   
25%      5463.000000      84.600000      -2.060000       5.600000   
50%     11361.000000      89.900000      -1.420000       5.890000   
75%     17366.000000      93.700000       1.250000       6.180000   
max     25195.000000     103.400000       4.610000       7.930000   

              batter        pitcher  spin_dir  spin_rate_deprecated  \
count  720637.000000  720637.000000       0.0                   0.0   
mean   599201.254543  599475.369626       NaN                   NaN   
std     62794.921518   63890.169024       NaN                   NaN   
min    405395.000000  424144.000000       NaN                   NaN   
25%    570256.000000  571578.000000       NaN                   NaN   
50%    608070.000000  608344.000000       NaN                   NaN   
75%    656305.000000  656302.000000       NaN                   NaN   
max    685503.000000  685503.000000       NaN                   NaN   

       break_angle_deprecated  break_length_deprecated           zone  \
count                     0.0                      0.0  720269.000000   
mean                      NaN                      NaN       9.091049   
std                       NaN                      NaN       4.221432   
min                       NaN                      NaN       1.000000   
25%                       NaN                      NaN       5.000000   
50%                       NaN                      NaN      11.000000   
75%                       NaN                      NaN      13.000000   
max                       NaN                      NaN      14.000000   

       hit_location          balls        strikes  game_year          pfx_x  \
count  160123.00000  720637.000000  720637.000000   720637.0  720268.000000   
mean        5.04890       0.883330       0.895841     2021.0      -0.106998   
std         2.63121       0.968423       0.828442        0.0       0.866944   
min         1.00000       0.000000       0.000000     2021.0      -2.560000   
25%         2.00000       0.000000       0.000000     2021.0      -0.840000   
50%         5.00000       1.000000       1.000000     2021.0      -0.180000   
75%         7.00000       2.000000       2.000000     2021.0       0.590000   
max         9.00000       4.000000       2.000000     2021.0       2.840000   

               pfx_z        plate_x        plate_z          on_3b  \
count  720269.000000  720269.000000  720269.000000   67695.000000   
mean        0.655249       0.044113       2.274805  600644.939168   
std         0.749900       0.846251       0.985328   61606.733375   
min        -2.130000      -6.100000      -5.070000  405395.000000   
25%         0.190000      -0.530000       1.640000  570731.000000   
50%         0.760000       0.040000       2.280000  608336.000000   
75%         1.280000       0.620000       2.920000  656537.000000   
max         3.040000       9.110000       9.390000  683734.000000   

               on_2b          on_1b   outs_when_up         inning  \
count  136361.000000  222451.000000  720637.000000  720637.000000   
mean   600228.028249  598740.188797       0.977346       4.922555   
std     62041.165331   63112.942277       0.818433       2.587854   
min    405395.000000  405395.000000       0.000000       1.000000   
25%    570731.000000  553902.000000       0.000000       3.000000   
50%    608336.000000  608070.000000       1.000000       5.000000   
75%    656555.000000  656305.000000       2.000000       7.000000   
max    683734.000000  683734.000000       2.000000      16.000000   

                hc_x           hc_y  tfs_deprecated  tfs_zulu_deprecated  \
count  123459.000000  123459.000000             0.0            

We have a lot of Null values and columns that don't really help in determining if a pitch is a strike or a ball so we need to drop the unnecessary columns and assess what is important.

---

### Assessing Nulls/NaNs

Since there are a total of 720,637 total pitches thrown in-game during the 2021 MLB season, we are gonna drop columns that are:
    
- Above 50% nulls
- Deprecated columns
- Not related to a pitch's mechanics, physics, etc. This includes:
    - A batter's influence on a pitch
    - Logisitics of the game
- Values related to expected run scores, hit expectancy and other values that don't determine pitch physics
        
In regards to columns that have null values, they will be assessed individually to determine how important they are for determing strikes vs ball calls.

In [10]:
mlb2021.columns

Index(['index', 'pitch_type', 'game_date', 'release_speed', 'release_pos_x',
       'release_pos_z', 'player_name', 'batter', 'pitcher', 'events',
       'description', 'spin_dir', 'spin_rate_deprecated',
       'break_angle_deprecated', 'break_length_deprecated', 'zone', 'des',
       'game_type', 'stand', 'p_throws', 'home_team', 'away_team', 'type',
       'hit_location', 'bb_type', 'balls', 'strikes', 'game_year', 'pfx_x',
       'pfx_z', 'plate_x', 'plate_z', 'on_3b', 'on_2b', 'on_1b',
       'outs_when_up', 'inning', 'inning_topbot', 'hc_x', 'hc_y',
       'tfs_deprecated', 'tfs_zulu_deprecated', 'fielder_2', 'umpire', 'sv_id',
       'vx0', 'vy0', 'vz0', 'ax', 'ay', 'az', 'sz_top', 'sz_bot',
       'hit_distance_sc', 'launch_speed', 'launch_angle', 'effective_speed',
       'release_spin_rate', 'release_extension', 'game_pk', 'pitcher.1',
       'fielder_2.1', 'fielder_3', 'fielder_4', 'fielder_5', 'fielder_6',
       'fielder_7', 'fielder_8', 'fielder_9', 'release_pos_y',
     

In [11]:
# Dropping Columns

mlb2021.drop(columns=['events','spin_dir','spin_rate_deprecated','break_angle_deprecated',
                     'break_length_deprecated','game_type','home_team','away_team',
                     'hit_location','game_year','on_3b','on_2b','on_1b',
                     'outs_when_up','inning','inning_topbot','hc_x','hc_y',
                     'tfs_deprecated','tfs_zulu_deprecated','umpire','sv_id',
                     'hit_distance_sc','launch_speed','launch_angle',
                     'game_pk', 'pitcher.1','fielder_2.1', 'fielder_3', 'fielder_4', 'fielder_5', 
                     'fielder_6', 'fielder_7', 'fielder_8', 'fielder_9',
                     'estimated_ba_using_speedangle', 'estimated_woba_using_speedangle',
                     'woba_value', 'woba_denom', 'babip_value', 'iso_value',
                     'launch_speed_angle', 'at_bat_number', 'pitch_number',
                     'home_score', 'away_score', 'bat_score', 'fld_score', 'post_away_score',
                     'post_home_score', 'post_bat_score', 'post_fld_score',
                     'if_fielding_alignment', 'of_fielding_alignment', 'spin_axis',
                     'delta_home_win_exp', 'delta_run_exp'], inplace=True)

In [12]:
# Assess what to do with columns that still have clean up
mlb2021.isnull().sum()

index                     0
pitch_type              365
game_date                 0
release_speed           401
release_pos_x           578
release_pos_z           578
player_name               0
batter                    0
pitcher                   0
description               0
zone                    368
des                       0
stand                     0
p_throws                  0
type                      0
bb_type              597099
balls                     0
strikes                   0
pfx_x                   369
pfx_z                   368
plate_x                 368
plate_z                 368
fielder_2                45
vx0                     368
vy0                     368
vz0                     368
ax                      368
ay                      368
az                      368
sz_top                  368
sz_bot                  368
effective_speed         609
release_spin_rate      3033
release_extension      2956
release_pos_y           578
pitch_name          

## Selecting what we need...

There is a column called `bb_type` where it means:

    Batted ball type, ground_ball, line_drive, fly_ball, popup.
    
Since this is referring to balls that were hit meaning there was no call about them being strikes or balls, we can omit the rows that have these filled to reduce the number of rows by about 20% (597,099/720637). We want the Null rows because there was a strike or ball call made, fouls are not accounted for but there is another column we can assess for that.

In [17]:
mlb2021 = mlb2021[mlb2021['bb_type'].isnull()]

Since not everything is based on the umpire's decision for a call, such as a foul or swinging strike, we are only going to be focused on `description` that are solely based on the umpire's call. We are going to focus on ball and called strikes only, unless one of the other descriptors is the same as either "called ball" or "called strike".

For clarification

- [swinging strikes and foul tips are the considered the same](https://www.mlb.com/glossary/rules/foul-tip)
- [swinging strike blocked is a swinging strike](https://www.mlb.com/braves/video/bruce-zimmermann-swinging-strike-blocked-to-j-d-martinez)
- hit by pitch could be either a ball or strike but since it hit the batter it will not be counted
- all bunts are considered to be strikes or fouls, the batter attempted to make contact resulting in a strike, not like a foul tip
- hit into play is basically means the batter made contact and the ball was thrown to get the batter out
- pitchouts are intentional balls in order to try and get the base runners stealing a base; since it's intentional we shall exclude them from being a "called ball"

In [20]:
mlb2021['description'].value_counts()

ball                       240415
foul                       127976
called_strike              118325
swinging_strike             76984
blocked_ball                17583
foul_tip                     6842
swinging_strike_blocked      4815
hit_by_pitch                 2137
foul_bunt                    1578
missed_bunt                   364
bunt_foul_tip                  41
pitchout                       34
hit_into_play                   4
foul_pitchout                   1
Name: description, dtype: int64

In [24]:
# We are selecting only "balls" and "called strikes"

mlb2021 = mlb2021.loc[((mlb2021['description'] == 'ball') | (mlb2021['description'] == 'called_strike'))]

In [25]:
mlb2021.isnull().sum()

index                     0
pitch_type              188
game_date                 0
release_speed           197
release_pos_x           293
release_pos_z           293
player_name               0
batter                    0
pitcher                   0
description               0
zone                    189
des                       0
stand                     0
p_throws                  0
type                      0
bb_type              358740
balls                     0
strikes                   0
pfx_x                   190
pfx_z                   189
plate_x                 189
plate_z                 189
fielder_2                23
vx0                     189
vy0                     189
vz0                     189
ax                      189
ay                      189
az                      189
sz_top                  189
sz_bot                  189
effective_speed         299
release_spin_rate      1567
release_extension      1537
release_pos_y           293
pitch_name          

After selecting the rows that we deemed necessary for pitch classification, we are left with a few hundred rows containing nulls, since the number is significantly smaller than what is remaining after the cleaning so far (358,740 rows), we are just going to drop the `bb_type` column and any rows containing nulls in any other column.

In [26]:
mlb2021.describe()

index  release_speed  release_pos_x  release_pos_z  \
count  358740.000000  358543.000000  358447.000000  358447.000000   
mean    11540.841696      88.854171      -0.708201       5.853890   
std      6873.348822       6.145198       1.891408       0.532926   
min         0.000000      30.400000      -4.640000       1.020000   
25%      5545.000000      84.600000      -2.060000       5.600000   
50%     11433.000000      90.100000      -1.430000       5.890000   
75%     17441.000000      93.700000       1.250000       6.180000   
max     25195.000000     102.800000       4.610000       7.930000   

              batter        pitcher           zone          balls  \
count  358740.000000  358740.000000  358551.000000  358740.000000   
mean   599201.257652  599537.159062      10.405362       0.734947   
std     63007.315559   63954.779138       3.798412       0.920400   
min    405395.000000  424144.000000       1.000000       0.000000   
25%    570256.000000  571578.000000       8.000000       0.000000   
50%    608324.000000  608344.000000      12.000000       0.000000   
75%    656371.000000  656322.000000      13.000000       1.000000   
max    685503.000000  685503.000000      14.000000       4.000000   

             strikes          pfx_x          pfx_z        plate_x  \
count  358740.000000  358550.000000  358551.000000  358551.000000   
mean        0.698676      -0.105212       0.662264       0.052523   
std         0.797260       0.872371       0.760882       1.000161   
min         0.000000      -2.560000      -2.130000      -6.100000   
25%         0.000000      -0.850000       0.200000      -0.710000   
50%         0.000000      -0.180000       0.790000       0.070000   
75%         1.000000       0.600000       1.290000       0.810000   
max         2.000000       2.840000       3.040000       9.110000   

             plate_z      fielder_2            vx0            vy0  \
count  358551.000000  358717.000000  358551.000000  358551.000000   
mean        2.302656  580528.072177       2.142379    -129.210981   
std         1.109283   67542.889077       6.100224       8.889585   
min        -4.980000  425772.000000     -18.902697    -149.318469   
25%         1.510000  543510.000000      -2.622294    -136.236486   
50%         2.240000  595978.000000       3.615375    -131.072517   
75%         3.070000  624513.000000       6.703079    -123.130251   
max         9.390000  680777.000000      20.914518     -40.960937   

                 vz0             ax             ay             az  \
count  358551.000000  358551.000000  358551.000000  358551.000000   
mean       -4.059020      -2.017115      26.661348     -23.195176   
std         3.478183      10.660861       3.979958       9.163859   
min       -20.138144     -35.092977      -2.089829     -50.549765   
25%        -6.523393     -10.972486      23.679222     -29.731756   
50%        -4.181910      -2.165570      26.683038     -22.368308   
75%        -1.773278       6.357195      29.638007     -15.301845   
max        15.486221      27.668111      44.536799      16.787233   

              sz_top         sz_bot  effective_speed  release_spin_rate  \
count  358551.000000  358551.000000    358441.000000      357173.000000   
mean        3.371688       1.584925        88.567036        2240.011098   
std         0.206090       0.112860         8.167960         330.073129   
min         2.500000       0.770000         0.000000          95.000000   
25%         3.240000       1.510000        84.500000        2079.000000   
50%         3.370000       1.590000        90.100000        2256.000000   
75%         3.500000       1.650000        93.800000        2431.000000   
max         4.470000       2.220000       103.900000        3598.000000   

       release_extension  release_pos_y  
count      357203.000000  358447.000000  
mean            6.298616      54.201223  
std             0.449637       0.448050  
min             2.800000      51.510000  
25%             6.0

In [27]:
mlb2021.drop(columns=['bb_type'], inplace=True)

mlb2021.dropna(inplace=True)

C:\Users\ncesm\anaconda3\envs\capstone\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(
<ipython-input-27-186b396925f0>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mlb2021.dropna(inplace=True)


In [28]:
mlb2021.describe()

index  release_speed  release_pos_x  release_pos_z  \
count  357046.000000   357046.00000  357046.000000  357046.000000   
mean    11545.508867       88.85640      -0.708696       5.854498   
std      6873.550643        6.14149       1.890495       0.532100   
min         0.000000       31.10000      -4.640000       1.020000   
25%      5552.000000       84.60000      -2.060000       5.600000   
50%     11441.000000       90.20000      -1.430000       5.890000   
75%     17445.000000       93.70000       1.240000       6.180000   
max     25195.000000      102.80000       4.610000       7.930000   

              batter        pitcher           zone          balls  \
count  357046.000000  357046.000000  357046.000000  357046.000000   
mean   599222.045148  599562.152697      10.403388       0.734838   
std     63005.146854   63930.026661       3.798998       0.920262   
min    405395.000000  424144.000000       1.000000       0.000000   
25%    570256.000000  571578.000000       8.000000       0.000000   
50%    608324.000000  608344.000000      12.000000       0.000000   
75%    656371.000000  656322.000000      13.000000       1.000000   
max    685503.000000  685503.000000      14.000000       4.000000   

             strikes          pfx_x          pfx_z        plate_x  \
count  357046.000000  357046.000000  357046.000000  357046.000000   
mean        0.698428      -0.105523       0.662639       0.052597   
std         0.797202       0.871983       0.760989       1.000125   
min         0.000000      -2.560000      -2.130000      -6.100000   
25%         0.000000      -0.850000       0.200000      -0.710000   
50%         0.000000      -0.180000       0.790000       0.070000   
75%         1.000000       0.600000       1.290000       0.810000   
max         2.000000       2.840000       2.580000       9.110000   

             plate_z      fielder_2            vx0            vy0  \
count  357046.000000  357046.000000  357046.000000  357046.000000   
mean        2.303670  580544.272427       2.144993    -129.214298   
std         1.108578   67543.254693       6.096522       8.884014   
min        -4.980000  425772.000000     -18.902697    -149.318469   
25%         1.510000  543510.000000      -2.613566    -136.239174   
50%         2.240000  595978.000000       3.615940    -131.075534   
75%         3.080000  624513.000000       6.702944    -123.132172   
max         9.390000  680777.000000      20.914518     -42.523580   

                 vz0             ax             ay             az  \
count  357046.000000  357046.000000  357046.000000  357046.000000   
mean       -4.060907      -2.021197      26.662392     -23.191499   
std         3.476971      10.654251       3.977550       9.164700   
min       -20.138144     -35.092977       2.473310     -50.549765   
25%        -6.524782     -10.969227      23.679844     -29.729233   
50%        -4.184034      -2.171779      26.683502     -22.361046   
75%        -1.775055       6.347693      29.638464     -15.297374   
max        15.486221      27.668111      44.536799      -1.322756   

              sz_top         sz_bot  effective_speed  release_spin_rate  \
count  357046.000000  357046.000000    357046.000000      357046.000000   
mean        3.371667       1.584948        88.874061        2240.012813   
std         0.206164       0.112884         6.297649         330.072530   
min         2.500000       0.770000        27.900000          95.000000   
25%         3.240000       1.510000        84.600000        2079.000000   
50%         3.370000       1.590000        90.100000        2256.000000   
75%         3.500000       1.650000        93.800000        2431.000000   
max         4.470000       2.220000       103.900000        3598.000000   

       release_extension  release_pos_y  
count      357046.000000  357046.000000  
mean            6.298698      54.201360  
std             0.449412       0.448513  
min             2.800000      51.510000  
25%             6.0

In [29]:
# Just by looking there is a change in numbers so we can easily assume
# that all nulls have been dropped but just to be sure we
# will run another

mlb2021.isnull().sum()

index                0
pitch_type           0
game_date            0
release_speed        0
release_pos_x        0
release_pos_z        0
player_name          0
batter               0
pitcher              0
description          0
zone                 0
des                  0
stand                0
p_throws             0
type                 0
balls                0
strikes              0
pfx_x                0
pfx_z                0
plate_x              0
plate_z              0
fielder_2            0
vx0                  0
vy0                  0
vz0                  0
ax                   0
ay                   0
az                   0
sz_top               0
sz_bot               0
effective_speed      0
release_spin_rate    0
release_extension    0
release_pos_y        0
pitch_name           0
dtype: int64

In [30]:
# mlb2021.to_csv('2021cleaned.csv', index=False) To avoid saving over it